keps/sig-auth/5284-constrained-impersonation/README.md

```diff
@@ -0,0 +1,1393 @@
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: ConstrainedImpersonation
+  - Components depending on the feature gate:
+    - kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+
+No. Impersonator with existing impersonate permission will still be allowed to impersonate
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. Set the feature gate to false and restart the kube-apiserver.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+No additional configuration is needed.
+
+###### Are there any tests for feature enablement/disablement?
+
+<!--
+The e2e framework does not currently support enabling or disabling feature
+gates. However, unit tests in each component dealing with managing data, created
+with and without the feature, are necessary. At the very least, think about
+conversion tests if API types are being modified.
+
+Additionally, for features that are introducing a new API field, unit tests that
+are exercising the `switch` of feature gate itself (what happens if I disable a
+feature gate after having objects written with the new field) are also critical.
+You can take a look at one potential example of such test in:
+https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
+-->
+
+Yes, this will be covered in the unit tests and integration tests.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+There is not impact on rollout, the impersonator with existing impersonate permission can still perform the action.
+When the system rollback, impersonator with `impersonate-on:` and `impersonate:` permission will no longer
+be authorized to impersonate. Impersonator will need to have the unscoped impersonate permission.
+
+
+###### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+Authz latency for impersonation is too long.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
```

Comment:
needs answer.  Without any persisted state, there is no lasting history to clear.